cmake/FindMKL.cmake

```diff
@@ -175,135 +127,183 @@ endif()
 
 if(WIN32)
     string(APPEND _mkl_libpath_suffix "_win")
+    set(_mkl_libname_prefix "")
+    set(_mkl_shared_lib "_dll.lib")
+    set(_mkl_static_lib ".lib")
 elseif(APPLE)
     string(APPEND _mkl_libpath_suffix "_mac")
-else()
+    set(_mkl_libname_prefix "lib")
+    set(_mkl_shared_lib ".dylib")
+    set(_mkl_static_lib ".a")
+else() # LINUX
     string(APPEND _mkl_libpath_suffix "_lin")
+    set(_mkl_libname_prefix "lib")
+    set(_mkl_shared_lib ".so")
+    set(_mkl_static_lib ".a")
 endif()
+set(_mkl_search_paths "${MKL_ROOT}"
+                      "${MKL_ROOT}/mkl"
+                      "${MKL_ROOT}/compiler")
 
-# Find MKL header
+# Functions: finds both static and shared MKL libraries
+#
+function(__mkl_find_library _varname _libname)
+    find_library(${_varname}_DYN
+          NAMES ${_mkl_libname_prefix}${_libname}${_mkl_shared_lib}
+          HINTS ${_mkl_search_paths}
+          PATH_SUFFIXES ${_mkl_libpath_suffix})
+    mark_as_advanced(${_varname}_DYN)
+    find_library(${_varname}_ST
+          NAMES ${_mkl_libname_prefix}${_libname}${_mkl_static_lib}
+          HINTS ${_mkl_search_paths}
+          PATH_SUFFIXES ${_mkl_libpath_suffix})
+    mark_as_advanced(${_varname}_ST)
+endfunction()
+
+# Find MKL headers
 #
 find_path(MKL_INCLUDE_DIR mkl.h
     HINTS ${MKL_ROOT}/include
-    )
+          ${MKL_ROOT}/mkl/include)
 mark_as_advanced(MKL_INCLUDE_DIR)
 
-# BLAS components (core MKL)
+# Group flags for static libraries on Linux (GNU, PGI, ICC -> same linker)
+#
+if(UNIX AND NOT APPLE)
+    set(_mkl_linker_pre_flags_ST "-Wl,--start-group")
+    set(_mkl_linker_post_flags_ST "-Wl,--end-group")
+endif()
+
+# Core MKL
 #
 __mkl_find_library(MKL_CORE_LIB mkl_core)
 
-__mkl_find_library(MKL_INTERFACE_32BIT_LIB mkl_intel_lp64)
-__mkl_find_library(MKL_INTERFACE_64BIT_LIB mkl_intel_ilp64)
+# Interface
+#
+__mkl_find_library(MKL_INTERFACE_INTEL_32BIT_LIB mkl_intel_lp64)
+__mkl_find_library(MKL_INTERFACE_INTEL_64BIT_LIB mkl_intel_ilp64)
+if(NOT APPLE AND CMAKE_Fortran_COMPILER_LOADED
+             AND CMAKE_Fortran_COMPILER_ID STREQUAL "GNU")
+    __mkl_find_library(MKL_INTERFACE_GF_32BIT_LIB mkl_gf_lp64)
+    __mkl_find_library(MKL_INTERFACE_GF_64BIT_LIB mkl_gf_ilp64)
+endif()
 
+# Threading
+#
 __mkl_find_library(MKL_SEQ_LIB mkl_sequential)
-if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND NOT APPLE)
+if(NOT APPLE AND (CMAKE_C_COMPILER_ID STREQUAL "GNU" OR
+                  CMAKE_CXX_COMPILER_ID STREQUAL "GNU" OR
+                  CMAKE_Fortran_COMPILER_ID STREQUAL "GNU"))
     __mkl_find_library(MKL_OMP_LIB mkl_gnu_thread)
 else()
     __mkl_find_library(MKL_OMP_LIB mkl_intel_thread)
 endif()
```

Comment:
I'm wondering about Clang openMP... Does anyone know if it is compatible with any of the MKL library?

Reply:
I am not sure... unless someone chimes in, I suggest we address it when/if we encounter issues.